operators/pkg/instautoctrl/terminator.go

```diff
@@ -0,0 +1,218 @@
+func (r *InstanceTerminationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if r.ReconcileDeferHook != nil {
+		defer r.ReconcileDeferHook()
+	}
+
+	log := ctrl.LoggerFrom(ctx, "instance", req.NamespacedName)
+
+	tracer := trace.New("reconcile", trace.Field{Key: "instance", Value: req.NamespacedName})
+	ctx = ctrl.LoggerInto(trace.ContextWithTrace(ctx, tracer), log)
+	defer tracer.LogIfLong(utils.LongThreshold())
+
+	// Get the instance object.
+	var instance clv1alpha2.Instance
+	if err := r.Get(ctx, req.NamespacedName, &instance); err != nil {
+		// TODO: need to requeue on error??
+		if !kerrors.IsNotFound(err) {
+			log.Error(err, "failed retrieving instance")
+		}
+		// Reconcile was triggered by a delete request.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Skip if the instance has not to be terminated.
+	if !utils.CheckSingleLabel(&instance, forge.InstanceTerminatorSelectorLabel, strconv.FormatBool(true)) {
+		return ctrl.Result{}, nil
+	}
+
+	// Check the selector label, in order to know whether to perform or not reconciliation.
+	if proceed, err := utils.CheckSelectorLabel(ctx, r.Client, instance.GetNamespace(), r.NamespaceWhitelist.MatchLabels); !proceed {
+		// If there was an error while checking, show the error and try again.
+		if err != nil {
+			log.Error(err, "failed checking selector labels")
+		}
+		return ctrl.Result{}, err
+	}
+
+	requeueResultOnError := ctrl.Result{RequeueAfter: InstanceStatusCheckRequeueTimeOnError}
+
+	// Check if the instance has to be terminated.
+	terminate, err := r.CheckInstanceTermination(ctx, &instance)
+	if err != nil {
+		log.Error(err, "failed checking instance termination")
+		return requeueResultOnError, err
+	}
+
+	if terminate {
+		if err := r.TerminateInstance(ctx, &instance); err != nil {
+			log.Error(err, "failed terminating instance")
+			return requeueResultOnError, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("requeueing instance")
+	return ctrl.Result{RequeueAfter: InstanceStatusCheckRequeueTime}, nil
+}
+
+// CheckInstanceTermination checks if the Instance has to be terminated.
+func (r *InstanceTerminationReconciler) CheckInstanceTermination(ctx context.Context, instance *clv1alpha2.Instance) (bool, error) {
+	if instance.Spec.CustomizationUrls == nil {
+		return false, errors.New("customization urls field is not set for Instance")
+	}
+
+	statusCheckURL := instance.Spec.CustomizationUrls.StatusCheck
+	if statusCheckURL == "" {
+		return false, errors.New("status check url field is not set for Instance")
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithName("statusCheck")
+	log.Info("performing instance status check")
+
+	statusCheckReponse := StatusCheckResponse{}
+	statusCode, err := utils.HTTPGetJSONIntoStruct(ctx, statusCheckURL, &statusCheckReponse)
+	if statusCode != http.StatusNotFound && err != nil {
+		return false, err
+	}
+
+	status := instance.Status
```

Comment:
```suggestion
	status := instance.Status.DeepCopy()
```

Better to be on the safe side